### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by Fatalll**

### CODE REVIEW

File: plugin/src/main/kotlin/tanvd/grazi/ide/ui/components/rules/GraziRulesPanel.kt [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            if (meta != null) onSelectionChanged(meta)
+        }
+    }
+
+    val panel by lazy {
+        panel {
+            panel(constraint = BorderLayout.NORTH) {
+                border = JBUI.Borders.emptyBottom(2)
+
+                with(DefaultActionGroup()) {
```
Comment (TanVD) [inline plugin/src/main/kotlin/tanvd/grazi/ide/ui/components/rules/GraziRulesPanel.kt]: `actionGroup { ...}`?

File: plugin/src/main/kotlin/tanvd/grazi/ide/ui/components/rules/GraziRulesTree.kt [hunk 2]
```diff
@@ -1,161 +1,51 @@
...
+    private val langs = HashSet(GraziConfig.get().enabledLanguages)
+    private val state = HashMap<String, RuleWithLang>()
 
     val isModified: Boolean
-        get() = state.values.any { it.lang in langs }
-
-    init {
-        _filter?.reset()
-    }
+        get() = state.values.onEach { println(it.rule.description) }. any { it.lang in langs }
```
Comment (TanVD) [inline plugin/src/main/kotlin/tanvd/grazi/ide/ui/components/rules/GraziRulesTree.kt]: `println`?

File: plugin/src/main/kotlin/tanvd/grazi/ide/ui/components/rules/GraziTreeNode.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                if (child is GraziTreeNode && child.resetMark(state)) {
+                    isChecked = true
+                }
+            }
+        }
+
+        return isChecked
+    }
+
+    override fun equals(other: Any?): Boolean {
```
Comment (TanVD) [inline plugin/src/main/kotlin/tanvd/grazi/ide/ui/components/rules/GraziTreeNode.kt:40]: Hm, do we really confident that there should be no hashcode?

File: plugin/src/test/kotlin/tanvd/grazi/grammar/LangDetectorTest.kt [hunk 4]
```diff
@@ -0,0 +1,202 @@
+package tanvd.grazi.grammar
+
+import tanvd.grazi.GraziTestBase
+import tanvd.grazi.language.Lang
+import tanvd.grazi.language.LangDetector
+
+class LangDetectorTest : GraziTestBase(true) {
```
Comment (TanVD) [inline plugin/src/test/kotlin/tanvd/grazi/grammar/LangDetectorTest.kt:6]: Let's think about adding such a test for markdown. To check that higlighting also works for all languages.

**Review by TanVD — CHANGES_REQUESTED** [review]

File: plugin/src/main/kotlin/tanvd/grazi/ide/ui/components/rules/GraziTreeNode.kt (hunk 3 above)
Comment (Fatalll) [reply plugin/src/main/kotlin/tanvd/grazi/ide/ui/components/rules/GraziTreeNode.kt:40]: Yeah, it breaks everything

File: plugin/src/main/kotlin/tanvd/grazi/ide/ui/components/rules/GraziRulesPanel.kt (hunk 1 above)
Comment (Fatalll) [reply plugin/src/main/kotlin/tanvd/grazi/ide/ui/components/rules/GraziRulesPanel.kt]: Done

File: plugin/src/main/kotlin/tanvd/grazi/ide/ui/components/rules/GraziRulesTree.kt (hunk 2 above)
Comment (Fatalll) [reply plugin/src/main/kotlin/tanvd/grazi/ide/ui/components/rules/GraziRulesTree.kt]: Removed

File: plugin/src/test/kotlin/tanvd/grazi/grammar/LangDetectorTest.kt (hunk 4 above)
Comment (Fatalll) [reply plugin/src/test/kotlin/tanvd/grazi/grammar/LangDetectorTest.kt:6]: Added

File: plugin/src/test/resources/ide/language/xml/Example.html [hunk 5]
```diff
@@ -5,8 +5,8 @@
     <title><warning descr="MORFOLOGIK_RULE_EN_US">Ttle</warning></title>
 </head>
 <body>
-<H1>Example <warning descr="MORFOLOGIK_RULE_EN_US">heder</warning> with <warning descr="EN_A_VS_AN">a</warning> error</H1>\
-<tag descr="it <warning descr="IT_VBZ">are</warning> error in tag <warning descr="MORFOLOGIK_RULE_EN_US">atribute</warning>"></tag>
+<H1>Example <warning descr="MORFOLOGIK_RULE_EN_US">hedder</warning> with <warning descr="EN_A_VS_AN">a</warning> error</H1>\
```
Comment (TanVD) [inline plugin/src/test/resources/ide/language/xml/Example.html:8]: Hm?

**Review by TanVD — CHANGES_REQUESTED** [review]

File: plugin/src/test/resources/ide/language/xml/Example.html (hunk 5 above)
Comment (Fatalll) [reply plugin/src/test/resources/ide/language/xml/Example.html:8]: We enabling 16 languages now, so some words with typos in tests exist in this languages :)

**Pull request merged**